core/src/hlist.rs

```diff
@@ -1108,40 +1108,81 @@ impl<F, FolderHeadR, FolderTail, H, Tail, Init> HFoldRightable<HCons<F, FolderTa
     for HCons<H, Tail>
 where
     Tail: HFoldRightable<FolderTail, Init>,
-    F: FnOnce(H, <Tail as HFoldRightable<FolderTail, Init>>::Output) -> FolderHeadR,
+    F: FnOnce(<Tail as HFoldRightable<FolderTail, Init>>::Output, H) -> FolderHeadR,
 {
     type Output = FolderHeadR;
 
     fn foldr(self, folder: HCons<F, FolderTail>, init: Init) -> Self::Output {
         let folded_tail = self.tail.foldr(folder.tail, init);
-        (folder.head)(self.head, folded_tail)
+        (folder.head)(folded_tail, self.head)
     }
 }
 
-impl<'a, F, R, H, Tail, Init> HFoldRightable<&'a F, Init> for HCons<H, Tail>
+impl<F, R, H, Tail, Init> HFoldRightable<F, Init> for HCons<H, Tail>
 where
-    Tail: HFoldRightable<&'a F, Init>,
-    F: Fn(H, <Tail as HFoldRightable<&'a F, Init>>::Output) -> R,
+    Tail: fn_foldr::FnHFoldRightable<F, Init>,
+    F: Fn(<Tail as HFoldRightable<F, Init>>::Output, H) -> R,
 {
     type Output = R;
 
-    fn foldr(self, folder: &'a F, init: Init) -> Self::Output {
-        let folded_tail = self.tail.foldr(folder, init);
-        (folder)(self.head, folded_tail)
+    fn foldr(self, folder: F, init: Init) -> Self::Output {
+        fn_foldr::FnHFoldRightable::real_foldr(self, folder, init).0
+    }
+}
+
+/// [`HFoldRightable`] inner mechanics for folding with a single closure/function.
+pub mod fn_foldr {
+    use super::{HCons, HFoldRightable, HNil};
+
+    /// A real `foldr` for the folder that is a closure or a function.
+    ///
+    /// Due to `HList` being a recursive struct and not linear array,
+    /// the only way to fold it is recursive.
+    ///
+    /// However, there are differences in the `foldl` and `foldr` traversing
+    /// the `HList`:
+    ///
+    /// 1. `foldl` calls `folder(head)` and then passes the ownership
+    ///     of the folder to the next recursive call.
+    /// 2. `foldr` passes the ownership of the folder to the next recursive call,
+    ///     and then tries to call `folder(head)`; but the ownership is already gone!
+    ///
+    /// Thus, it should either take a reference to the folder or return ownership back.
+    /// This trait does the latter to avoid lifetimes in the trait definition.
+    pub trait FnHFoldRightable<Folder, Init>: HFoldRightable<Folder, Init> {
```

Comment:
One more unresolved question.
This could be used later for other right-folds with ownership problems. The name and docs are closure/function related, though, and due to module and trait being public, changing names later would be a breaking change.
So, yeah, the question is: any ideas for better generic name and docs for this thing?

Reply:
This is a good point... I think it's generally good to use a name that is descriptive wrt the purpose of the trait and easily understood. Just going to throw this out there: "ClosureUnifier"?? No worries if it's horrible, just wanted to get the ball rolling :)

Reply:
It's a tricky question.  HFoldRightRecoverable, since you recover the function?  I dunno.

Honestly though there already a lot of things frunk can't add without a semver breaking change, so it doesn't feel like a big deal to me. It's a wonder that frunk is only at 0.3, but iirc that's because it historically did not follow semver.

Reply:
Changed `fn_foldr` to `foldr_owned` and `FnHFoldRightable` to `HFoldRightableOwned`. Maybe not the clearest name, but it does decent job at describing the trait and is generic enough. If everyone's okay with the name, this conversation could be resolved, and the PR could be merged, I think.